File: leetcode/10-100/leetcode71.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <vector>
using namespace std;
class Solution {
private:
    vector<string> s;

public:
    string simplifyPath(string path)
    {
        // clear the stack
        s.clear();

        int i = 0;
        string str;
        for (;;) {
            str = getDirName(path, i);
            if (str == "")
                break;
            if (str == "..") {
                if (!s.empty())
                    s.pop_back();
            } else if (str == ".") {
                continue;
            } else {
                s.push_back(str);
            }
        }

        str = "";
        for (string& t : s) {
            str += "/" + t;
        }
        if (str == "")
            str = "/";
        return str;
    }

    string getDirName(string path, int& i)
    {
        int start = i;
        while (start < path.length()) {
            if (path[start] != '/')
                break;
            start++;
        }
        i = start;
        while (i < path.length()) {
            if (path[i] == '/')
                break;
            i++;
        }
        if (i == start)
            return "";
        return path.substr(start, i - start);
    }
};
int main(void) { return 0; }

File: leetcode/10-99/leetcode71.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        p = []
        while path:
            path = self.eatSlash(path)
            if not path:
                break
            dirName = self.getDirName(path)
            if dirName == '..':
                if p:
                    p.pop()
            elif dirName == '.':
                pass
            else:
                p.append(dirName)
            path = path[len(dirName):]
        
        return "/" + "/".join(p)

    def getDirName(self, s:str) -> str:
        res = []
        for c in s:
            if c == '/':
                break
            res.append(c)
        return "".join(res)
    def eatSlash(self, s: str) -> str:
        for i in range(len(s)):
            if s[i] != '/':
                return s[i:]
        return ""
